### DISCUSSION TITLE
Can `quarto-pub` be created as a new branch similar to `gh-pages`?

### DISCUSSION THREAD

**Discussion by batmanscode** [quarto-dev/quarto-cli#3729] [Q&A] [answered]
I'd like the main branch not to have `_docs` and `_proc` in commit history

**Comment by mcanouil** [comment, marked as answer]
I am not sure to understand, `quarto-pub` is a host where `gh-pages` is a GitHub feature which has different setup, including one to publish from a Git branch.

This being said, `_docs` and `_proc` are part of your history because you added those directories.
Simply add those directories in `.gitignore`.

**Reply by batmanscode** [reply]
Sorry, I could've been more clear. Thanks for the reply!

I mean is there a way to have the docs built and published with actions alone?

At the moment, I have to include _procs and _docs in my repo for the quarto-pub action to work

**Reply by mcanouil** [reply]
You probably did not set properly (as per your goal) the GitHub Action workflow.
Could you share it, or share the repository?

**Reply by batmanscode** [reply]
Probably not haha. I started with the docs, then changed a bunch to experiment and now back to the version in the docs.

Unfortunately I can't share the repo at the moment since it's private but I can certainly share the workflow!

I'm using this with nbdev, which is why I tried to use quarto-pub at all btw! Since nbdev has a workflow built in to create docs in a new branch using GitHub pages but that feature only works seamlessly with public repos.

The relevant nbdev docs that create `_docs` and `_proc`: https://nbdev.fast.ai/explanations/docs.html#deploying-your-docs-on-other-platforms


Workflow:
```
on: 
   workflow_dispatch: 
   push: 
     branches: nbdev-setup 
  
 name: Quarto Publish 
  
 jobs: 
   build-deploy: 
     runs-on: ubuntu-latest 
     steps: 
       - name: Check out repository 
         uses: actions/checkout@v2  
  
       - name: Set up Quarto 
         uses: quarto-dev/quarto-actions/setup@v2 
          
       - name: Install Python and Dependencies 
         uses: actions/setup-python@v4 
         with: 
           python-version: '3.8' 
           cache: 'pip'        - run: pip install -r requirements.txt
        
       - name: Render and Publish  
         uses: quarto-dev/quarto-actions/publish@v2 
         with: 
           target: quarto-pub 
           path: _proc 
           QUARTO_PUB_AUTH_TOKEN: ${{ secrets.QUARTO_PUB_AUTH_TOKEN }}
```

**Reply by mcanouil** [reply]
Sorry for the late reply, you need to use a different GitHub Action instead of the current "Render and Publish" step, _i.e._, you need to use for example <https://github.com/quarto-dev/quarto-actions/tree/main/render> followed by a step to create and push to the desired branch.

For example (not tested)

```
- name: Render and Publish  
  run: |
    quarto render --output-dir <path/to/file/folders/on/main/branch>
    git checkout --orphan quarto-pub
    git checkout main -- <path/to/file/folders/on/main/branch>
    git commit -m 'ci: rebuild website' || echo "No changes to commit"
    git push origin || echo "No changes to commit"  
```

**Reply by batmanscode** [reply]
> Sorry for the late reply, you need to use a different GitHub Action instead of the current "Render and Publish" step, _i.e._, you need to use for example <https://github.com/quarto-dev/quarto-actions/tree/main/render> followed by a step to create and push to the desired branch.
> 
> For example (not tested)
> 
> ```
> - name: Render and Publish  
>   run: |
>     quarto render --output-dir <path/to/file/folders/on/main/branch>
>     git checkout --orphan quarto-pub
>     git checkout main -- <path/to/file/folders/on/main/branch>
>     git commit -m 'ci: rebuild website' || echo "No changes to commit"
>     git push origin || echo "No changes to commit"  
> ```
> 

Not a problem! This is exactly what I was looking for, thank you 😃

**Comment by cderv** [comment]
The `quarto publish` command can be used to publish to GH PAGES (https://quarto.org/docs/publishing/github-pages.html). You need to do the correct setup for the `gh-pages` branch instead of the folder in default branch. 

You can do it from local environment, or from Github actions - docs are https://github.com/quarto-dev/quarto-actions/blob/main/publish/README.md#github-pages and https://quarto.org/docs/publishing/github-pages.html#github-action

Did you try any of this as we document ? Is this not working for you as expected ?

**Reply by mcanouil** [reply]
Here the idea was to use a other branch or still there is some confusion ...

The GHA workflow includes `target: quarto-pub` as target of the `publish` Quarto GHA, see <https://github.com/quarto-dev/quarto-cli/discussions/3729#discussioncomment-4471630>.

**Reply by cderv** [reply]
Oh ok sorry - it seems I misunderstood the whole issue here... Thanks for correcting me! 

Note that a `checkout` actions can be used several time in a repo to checkout another repo or another branch, this could be used to checkout the specific branch and then call the publish quarto action on this new checkout folder probably. 

Anyway, thank for correcting me. I'll hide this post above.

**Comment by batmanscode** [comment]
> Note that a `checkout` actions can be used several time in a repo to checkout another repo or another branch, this could be used to checkout the specific branch and then call the publish quarto action on this new checkout folder probably
> 
> 

Useful, thanks @cderv 😃

And sorry, I forgot to label as answered 😅, thanks for adding that in

I can't reply in that comment thread anymore so doing it here. For more context, this was about publishing docs via nbdev. Usually it happens with gh-pages (in a separate branch) but the repo being worked on was private so couldn't use the default nbdev workflow

I wanted to have a similar set up, where a new branch is created just like gh-pages, but using quarto-pub instead. The reason being that it's a private repo

So private repo -> create new branch for docs -> publish on quarto from the dedicated docs branch (instead of gh-pages, which is the default behaviour for nbdev)

**Reply by cderv** [reply]
Thanks for the additional information ! 

> I can't reply in that comment thread anymore so doing it here.

I see that now 🤦‍♂️  sorry
